frcxx_nft/src/state.rs

```diff
@@ -171,109 +221,1284 @@ impl NFTState {
+        let new_owner_data = match owner_map.get(&new_owner_key)? {
+            Some(data) => OwnerData { balance: data.balance + 1, ..data.clone() },
+            None => OwnerData { balance: 1, operators: vec![] },
+        };
+        owner_map.set(new_owner_key, new_owner_data)?;
+
         Ok(())
     }
+
+    /// Asserts that the actor owns the token and returns a copy of the TokenData
+    ///
+    /// Returns TokenNotFound if the token_id is invalid or NotOwner if the actor does not own
+    /// own the token.
+    pub fn owns_token<BS: Blockstore>(
+        token_array: &Amt<TokenData, &BS>,
+        actor: ActorID,
+        token_id: TokenID,
+    ) -> Result<TokenData> {
+        let token_data =
+            token_array.get(token_id)?.ok_or_else(|| StateError::TokenNotFound(token_id))?;
+        match token_data.owner == actor {
+            true => Ok(token_data.clone()),
+            false => Err(StateError::NotOwner { actor, token_id }),
+        }
+    }
+
+    /// Checks whether an operator is approved to transfer/burn a token
+    pub fn approved_for_token<BS: Blockstore>(
+        token_array: &Amt<TokenData, &BS>,
+        owner_map: &Hamt<&BS, OwnerData>,
+        operator: ActorID,
+        token_id: TokenID,
+    ) -> Result<bool> {
+        let token_data = token_array
+            .get(token_id)?
+            .ok_or_else(|| StateError::InvariantFailed(format!("token {token_id} not found")))?;
+
+        // operator is approved at token-level
+        if token_data.operators.contains_actor(&operator) {
+            return Ok(true);
+        }
+
+        // operator is approved at account-level
+        let owner_account = owner_map.get(&actor_id_key(token_data.owner))?.ok_or_else(|| {
+            StateError::InvariantFailed(format!("owner of token {token_id} not found"))
+        })?;
+        if owner_account.operators.contains_actor(&operator) {
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+
+    /**
+     * Converts a MintIntermediate to a MintReturn
+     *
+     * This function should be called on a freshly loaded or known-up-to-date state
+     */
+    pub fn mint_return<BS: Blockstore>(
+        &self,
+        bs: &BS,
+        intermediate: MintIntermediate,
+    ) -> Result<MintReturn> {
+        let balance = self.get_balance(bs, intermediate.to)?;
+
+        Ok(MintReturn {
+            balance,
+            supply: self.total_supply,
+            token_ids: intermediate.token_ids,
+            recipient_data: intermediate.recipient_data,
+        })
+    }
+
+    /**
+     * Converts a TransferIntermediate to a TransferReturn
+     *
+     * This function should be called on a freshly loaded or known-up-to-date state
+     */
+    pub fn transfer_return<BS: Blockstore>(
+        &self,
+        bs: &BS,
+        intermediate: TransferIntermediate,
+    ) -> Result<TransferReturn> {
+        // TODO: optimise a pattern to avoid reading the owner data hamt twice
```

Comment:
There's nothing easy here. The HAMT library could be extended with a batch get, but there's not much you can do about it from here at the moment.